sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ReviewSqlQueriesForSecurityVulnerabilities.cs

```diff
@@ -85,11 +114,90 @@ protected override void Initialize(SonarAnalysisContext context)
+
+                    cc.RegisterSyntaxNodeActionInNonGenerated(
+                        c => WriteUCFG<MethodDeclarationSyntax>(c, x => (CSharpSyntaxNode)x.Body ?? x.ExpressionBody?.Expression),
+                        SyntaxKind.MethodDeclaration);
+
+                    cc.RegisterSyntaxNodeActionInNonGenerated(
+                        c => WriteUCFG<AccessorDeclarationSyntax>(c, node => node.Body),
+                        SyntaxKind.GetAccessorDeclaration,
+                        SyntaxKind.SetAccessorDeclaration);
+
+                    cc.RegisterSyntaxNodeActionInNonGenerated(
+                        c => WriteUCFG<PropertyDeclarationSyntax>(c, node => node.ExpressionBody?.Expression),
+                        SyntaxKind.PropertyDeclaration);
+                });
+        }
+
+        private void WriteUCFG<TDeclarationSyntax>(SyntaxNodeAnalysisContext context, Func<TDeclarationSyntax, CSharpSyntaxNode> getBody)
+            where TDeclarationSyntax : SyntaxNode
+        {
+            var declaration = (TDeclarationSyntax)context.Node;
+
+            var symbol = context.SemanticModel.GetDeclaredSymbol(declaration);
+
+            if (symbol == null ||
+                symbol.IsAbstract ||
+                symbol.IsExtern ||
+                !CSharpControlFlowGraph.TryGet(getBody(declaration), context.SemanticModel, out var cfg))
+            {
+                return;
+            }
+
+            var ucfg = new UCFG
+            {
+                MethodId = MethodIdProvider.Create(symbol),
+                Location = UniversalControlFlowGraphBuilder.GetLocation(declaration),
+            };
+
+            ucfg.BasicBlocks.AddRange(new UniversalControlFlowGraphBuilder(context.SemanticModel, cfg).Build());
+            ucfg.Parameters.AddRange(symbol.GetParameters().Select(p => p.Name));
+
+            var path = Path.Combine(protobufDirectory, $"ucfg_{configurationName}_{Interlocked.Increment(ref protobufFileIndex)}.pb");
+            using (var stream = File.Create(path))
+            {
+                ucfg.WriteTo(stream);
+            }
         }
 
         private static bool IsSanitizedQuery(ExpressionSyntax expression, SemanticModel model)
         {
             return model.GetConstantValue(expression).HasValue;
         }
+
+        private bool TryReadConfiguration(AnalyzerOptions options)
+        {
+            var basePath = configuration.GetProjectOutput(options);
+
+            // the current compilation output dir - "<root>/.sonarqube/out/<index>" where index is 0, 1, 2, etc.
+            if (basePath != null)
+            {
+                // "<root>/.sonarqube/out/0" -> "0" etc.
+                configurationName = Path.GetFileName(basePath);
+
+                // "<root>/.sonarqube/out/0" -> "<root>/.sonarqube/out/ucfg_cs"
+                protobufDirectory = Path.Combine(Path.GetDirectoryName(basePath), $"ucfg_{AnalyzerLanguage.CSharp}");
+
+                Directory.CreateDirectory(protobufDirectory);
```

Comment:
Might be worth asserting to check the assumption that the directory doesn't already exist.